Deletes init-time tick overruns

Cyberboss opened the pull request:
Or minimizes them, rather

## About The Pull Request

Used tracy to profile world init and addressed all the major causes of tick overrun. Before we had ticks running up to ~5-7s. Now we're down to ~340ms tops.
![image](https://user-images.githubusercontent.com/8171642/234470126-524ea1ee-bd2d-45b4-ae40-7d62977d9301.png)

## Why It's Good For The Game

Server responsiveness is always good.

:cl:
code: Improved server responsiveness during world initialization.
tweak: Character select screen will be unavailable until the Early Assets subsystem initializes if the server's cache isn't warmed up. To offset this, it now initializes before the map is loaded.
spriteadd: New sprite for character select screen enabling, credit to Kryson.
fix: Fixed spritesheet caching sometimes erroneously missing.
/:cl:

- Added CHECK_TICKs to spritesheet inserting and some overwhelming create_spritesheets procs/subprocs. No adverse effects that I can see.
- Added a CHECK_TICKed version of cannonize_contained_turfs for /area/space/ to run at the end of mapping init.
- Merged SSasset_loading (which was SS_NO_INIT) into SSassets (which was SS_NO_FIRE).
- Tried to make clear the difference between the states of asset datum loading (unregistered, registered, generated).
- The calls to LateInitialize from SSatoms/Initialize are now CHECK_TICKed as they should have been years ago.
- Disabled the character select button until SSearly_assets Initializes. Ideally this would lazy load but that requires a tangential refactor to how tgui requests assets. Will make that PR later
- New dev config to disable early asset loading so this change doesn't hang init times for all coders.
- Fixed lobby buttons not having INITIALIZE_IMMEDIATE.
- Added signals for when prefs assets are generated and when a single asset is generated.
- Improved yielding in realize_spritesheets.
- /datum/asset_transport/register_assets is now CHECK_TICKed. Added proc/register_asset_immediate for cases where this is not functional due to SHOULD_NOT_SLEEP (icon2html())
- Moved SSearly_assets init to before SSmapping
- Stopped CACHE_ASSETS from forcing spritesheets to generate immediately on a cache miss.
- Remove unused list SSassets.preload.
- Fixed incorrect spritesheet cache misses.

https://user-images.githubusercontent.com/8171642/235014235-d6c55b7b-e59e-48a7-920e-63987d01ac41.mp4

Cyberboss committed:
Improve all outstanding init time tick overruns

Cyberboss commented on code/game/area/areas.dm:
```diff
@@ -230,6 +230,25 @@ GLOBAL_LIST_EMPTY(teleportlocs)
 	contained_turfs -= turfs_to_uncontain
 	turfs_to_uncontain = list()
 
+/area/space/proc/cannonize_contained_turfs_slow()
+	// space has a gigantic fucking amount of turfs after loading the map
+	// Since we know it's safe to do so right now, chunk up the uncontainment
```
Biggest gains right here. The first time /area/space called cannonize_contained_turfs() it took around 6s for the BYOND list `-=` operation

Mothblocks replied:
@LemonInTheDark

LemonInTheDark replied:
Copying off the core list seems like something that'd be pretty slow, tho I guess not cause it's not moving all that much at once?

LemonInTheDark replied:
/area/space should not have a large amount of uncontained turfs, we put in effort when loading large spaces to add them like... in place if that makes sense.
It's strange that that isn't the case. I'd be curious to know where the problem turfs came from. ruins maybe?

Cyberboss replied:
As far as speed goes, it's about the same amount of time minus the tick overruns.

Cyberboss commented on code/controllers/subsystem/atoms.dm:
```diff
@@ -52,6 +52,7 @@ SUBSYSTEM_DEF(atoms)
 			if(QDELETED(A))
 				continue
 			A.LateInitialize()
+			CHECK_TICK
```
I asked myself a billion times why this wasn't already the case. Another source of tick overrun was not checking between these calls.

Cyberboss commented on code/modules/asset_cache/transports/asset_transport.dm:
```diff
@@ -40,6 +40,8 @@
 	if (!istype(ACI))
 		ACI = new(asset_name, asset, file_hash, dmi_file_path)
+		// MD5 calls are slow
+		CHECK_TICK
```
Big asset init savings, please confirm sleeping is okay here.

optimumtact replied:
are you really just making changes without knowing?

Cyberboss replied:
I'm not familiar with the system but at a glance there's no reason sleeping should be a problem

Cyberboss commented on code/modules/asset_cache/asset_list.dm:
```diff
@@ -361,6 +361,7 @@ GLOBAL_LIST_EMPTY(asset_datums)
 		queuedInsert(sprite_name, I, icon_state, dir, frame, moving)
 	else
 		to_generate += list(args.Copy())
```
Bigger asset init savings, please confirm sleeping is okay here.

Mothblocks replied:
It isn't, from experience

Cyberboss replied:
Changed this particular case

Cyberboss commented:
Uhhh, what are the linters complaining about??

Mothblocks commented:
doesnt this DRAMATICALLY increase init time? every time i do one of these in hot paths it increases init time by like 6-15 seconds

Mothblocks commented on code/modules/client/preferences/ai_core_display.dm:
```diff
@@ -12,6 +12,7 @@
 
 	for (var/screen in GLOB.ai_core_display_screens - "Portrait" - "Random")
 		values[screen] = icon('icons/mob/silicon/ai.dmi', resolve_ai_icon_sync(screen))
+		CHECK_TICK
```
do not CHECK_TICK in a single thing in preferences or even assets as a whole.

i have tried three times. every time byond shits itself and gives completely useless and unsolvable icon errors. theres a very good reason it does what it does

LemonInTheDark replied:
this may have been reduced when I did the icon refreshing thing. Not sure how sleeping precisely here will play with that tho

Mothblocks commented on code/modules/client/preferences/assets.dm:
```diff
@@ -8,6 +8,7 @@
 	var/list/to_insert = list()
 
 	for (var/preference_key in GLOB.preference_entries_by_key)
+		CHECK_TICK
```
this 100% has given me huge shit before

Mothblocks commented on code/modules/client/preferences/clothing.dm:
```diff
@@ -21,6 +21,7 @@
 		icon_with_socks.Scale(32, 32)
 
 		values[accessory_name] = icon_with_socks
+		CHECK_TICK
```
this has also given me huge shit before. everything youre touching here

Mothblocks reviewed (dismissed):
Do not touch prefs or assets at all. EVERY time I have added something there to yield to MC it falls apart on prod specifically for BYOND reasons we never pinned down

Prefs assets specifically are cached on prod anyway (rotated every TGS update), so this would not have noticeable effects there

Mothblocks commented:
I will also need to check this for myself because I definitely do remember this *murdering* init times

Mothblocks commented:
Though that being said once the preferences stuff is un-CHECK_TICK'd, I'm curious if this holds true for anything else:

> Before we had ticks running up to ~5-7s.

Mothblocks commented:
Linter is complaining because signal handlers can't yield, but you added yielding to register_asset, which some stuff (mostly reagents, to show the bubbling effect in chat I think?) uses. Ideally those consumers would be using the 515 appearance stuff when we require it in use for clients

LemonInTheDark commented:
> Do not touch prefs or assets at all. EVERY time I have added something there to yield to MC it falls apart on prod specifically for BYOND reasons we never pinned down

I thought I fixed that, was an issue with icons going like, weird after a long enough period of time. fixed by creating a new icon with the old one as a base

Mothblocks commented:
Okay, you might be right.

I will see how much this impacts init times

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Cyberboss commented:
Init time with changes: 1:24
Without: 1:22

Completely negligible

Pic is without changes
![image](https://user-images.githubusercontent.com/8171642/234868199-b4b4919a-4161-4262-b81c-901e023db54a.png)

Cyberboss committed:
Alternate version of register_asset that executes without yielding for icon2html(). Fixes lint errors

Cyberboss commented:
My SpacemanDMM wasn't updating and wasn't catching some SHOULD_NOT_SLEEP execution paths. Fixed those.

A quick spot check shows nothing wrong with prefs.

If something shouldn't be sleeping it really should have the macro and why.

Cyberboss commented:
Holy fuck, no one realized this happens the first time someone tries to open the character select screen?

Cyberboss commented:
Okay 
1. Yielding in that call fucks the setup menu. No surprise there.
2. Again, if these procs should not sleep they should be marked as such
3. If the menu can't be opened due to a massive tick spike like this WE SHOULD NOT LET PLAYERS DO IT

Cyberboss commented:
I'll let the above go in this PR, it's a much bigger refactor to the asset subsystems. Will open another PR with those changes

Cyberboss committed:
Fixing opening character setup on a cache miss overrunning ticks by >15s

Also merges SSasset_loading into SSassets

Cyberboss committed:
SSresearch doesnt need to be initialized early, so it doesnt need early assets

Cyberboss committed:
Dev config option to disable early asset init which nerfs init times

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Cyberboss commented:
I lied, the change turned out to be inter-dependent. I've added it here. I've addressed things in a way so that dev init times don't increase and they'll only increase on prod if the prefs cache misses.

I'll update the desc with the full change list

Cyberboss committed:
Ignore /atom/movable/screen/lobby/button/character_setup in create_and_destroy tests as it now requires a client/prefs on init

Cyberboss committed:
This proc can be much simpler

Cyberboss committed:
Fix backwards naming for config entry

Cyberboss committed:
Add doc comment

Cyberboss committed:
Moves early asset init to before mapping

Cyberboss committed:
Fix fucky indentation

Cyberboss committed:
Add disabled Character Setup button sprites, credit to @Krysonism

Cyberboss committed:
Revert "SSresearch doesnt need to be initialized early, so it doesnt need early assets"

This reverts commit 9e26d5379157673ceb544937d845cf8edc4a7d52.

Cyberboss committed:
Allow prefs to load in dev mode

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss commented:
I think I orthogonally fixed the screenshot issue? I can't repro it now and I'm constantly re-running the tests to make sure it's not just flaky.

So this is ready for review.

Cyberboss commented:
It was probably caused by [9e26d53](https://github.com/tgstation/tgstation/pull/74986/commits/9e26d5379157673ceb544937d845cf8edc4a7d52), which I've reverted.

san7890 commented on code/_onclick/hud/new_player.dm:
```diff
@@ -11,19 +11,21 @@
 
+INITIALIZE_IMMEDIATE(/atom/movable/screen/lobby)
+
+/atom/movable/screen/lobby/Initialize(mapload, datum/hud/new_player/_hud)
+	. = ..()
+	hud = _hud
```
This should really be done like this:

```suggestion
/atom/movable/screen/lobby/Initialize(mapload, datum/hud/new_player/hud)
	. = ..()
	src.hud = hud
```

per https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar

san7890 commented on code/_onclick/hud/new_player.dm:
```diff
@@ -42,6 +44,10 @@
 	/// The ref of the mob that owns this button. Only the owner can click on it.
 	var/owner
 
+/atom/movable/screen/lobby/button/Initialize(mapload, datum/hud/new_player/_hud, mob/_owner)
+	. = ..()
+	owner = REF(_owner)
```
ditto on args

```suggestion
/atom/movable/screen/lobby/button/Initialize(mapload, datum/hud/new_player/hud, mob/owner)
	. = ..()
	src.owner = REF(owner)
```

san7890 reviewed (dismissed):
i only fixed the first two examples i saw, be sure to use `src.var = var` whenever applicable

Cyberboss commented on code/_onclick/hud/new_player.dm:
```diff
@@ -103,6 +109,20 @@
 	icon_state = "character_setup"
 	base_icon_state = "character_setup"
 
+/atom/movable/screen/lobby/button/character_setup/Initialize(mapload, datum/hud/new_player/_hud, mob/_owner)
+	. = ..()
+	var/datum/preferences/preferences = _owner.client.prefs
```
```suggestion
/atom/movable/screen/lobby/button/character_setup/Initialize(mapload, datum/hud/new_player/hud, mob/owner)
	. = ..()
	var/datum/preferences/preferences = owner.client.prefs
```

Cyberboss commented on code/_onclick/hud/new_player.dm:
```diff
@@ -121,7 +141,7 @@
 	var/ready = FALSE
 
-/atom/movable/screen/lobby/button/ready/Initialize(mapload)
+/atom/movable/screen/lobby/button/ready/Initialize(mapload, datum/hud/new_player/_hud, mob/_owner)
```
```suggestion
/atom/movable/screen/lobby/button/ready/Initialize(mapload, datum/hud/new_player/hud, mob/owner)
```

Cyberboss commented on code/_onclick/hud/new_player.dm:
```diff
@@ -166,7 +186,7 @@
 	enabled = FALSE
 
-/atom/movable/screen/lobby/button/join/Initialize(mapload)
+/atom/movable/screen/lobby/button/join/Initialize(mapload, datum/hud/new_player/_hud, mob/_owner)
```
```suggestion
/atom/movable/screen/lobby/button/join/Initialize(mapload, datum/hud/new_player/hud, mob/owner)
```

Cyberboss commented on code/_onclick/hud/new_player.dm:
```diff
@@ -236,7 +256,7 @@
 	enabled = FALSE
 
-/atom/movable/screen/lobby/button/observe/Initialize(mapload)
+/atom/movable/screen/lobby/button/observe/Initialize(mapload, datum/hud/new_player/_hud, mob/_owner)
```
```suggestion
/atom/movable/screen/lobby/button/observe/Initialize(mapload, datum/hud/new_player/hud, mob/owner)
```

Cyberboss committed:
Apply suggestions from code review

Co-authored-by: san7890 <the@san7890.com>

Mothblocks commented on code/__DEFINES/preferences.dm:
```diff
@@ -136,3 +136,6 @@
 #define PLAYTIME_VETERAN 300000 // 5,000 hours
+
+/// Signal sent when all ui assets for preferences are loaded
+#define COMSIG_PREFS_ASSETS_LOADED "preferences_assets_loaded"
```
Should be with the other signals

Mothblocks commented on code/__DEFINES/dcs/signals/signals_assets.dm:
```diff
@@ -0,0 +1,2 @@
+/// Sent when an asset datum is fully generated
```
Needs the same info as the rest of the signal comments, specifically what the `source` is (I don't care about the call source)

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Mothblocks commented on config/config.txt:
```diff
@@ -623,6 +623,12 @@ MOTD motd.txt
 ## but enabled on production (the default).
 CACHE_ASSETS 0
 
+## Loading early assets, primarily used by the preferences menu, greatly slows down init times
+## This dev option here is to bypass that init phase and let the preferences menu queue normally, or overrun the tick to load on demand
+## This should be disabled (through `ENABLE_EARLY_ASSETS 0`) on development,
+## but enabled on production (the default).
+ENABLE_EARLY_ASSETS 0
```
Can you explain the thought process of this a bit more?

Also this *needs* to be enabled in the CI config (I think `ciconfig.txt`?). The literal **only** point of early assets is to guarantee that preferences menu assets running before atoms is okay.

Cyberboss replied:
Explained in below comment, but, to reiterate, the only point of this is to keep the current behaviour of hiding the 12-14s additional init time unless it's needed by a dev (by clicking setup character).

Will add to CI config.

Mothblocks commented on code/modules/client/preferences.dm:
```diff
@@ -343,6 +345,40 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+	SIGNAL_HANDLER
+	UnregisterSignal(asset, COMSIG_ASSET_GENERATED)
+
+	if(--asset_loading_count == 0)
```
```suggestion
	asset_loading_count--
	if(asset_loading_count == 0)
```

We don't do this anymore

Mothblocks commented on code/modules/client/preferences/middleware/keybindings.dm:
```diff
@@ -81,6 +79,7 @@
 
 /datum/asset/json/keybindings
 	name = "keybindings"
+	early = TRUE
```
Why bother? It doesn't need to be.

Cyberboss replied:
It's required by the prefs menu, see my comment below

Mothblocks replied:
It's needed by prefs menu but it doesn't actually need to be in early assets because it doesn't do anything other than read a GLOB

Mothblocks commented on code/modules/client/preferences.dm:
```diff
@@ -192,13 +197,10 @@ GLOBAL_LIST_EMPTY(preferences_datums)
-		get_asset_datum(/datum/asset/spritesheet/preferences),
-		get_asset_datum(/datum/asset/json/preferences),
-	)
+	var/list/assets = ..()
```
There is no `..()` for `/datum/ui_assets`--we shouldn't do this because it makes it less obvious what is actually going on here

Mothblocks commented on code/modules/client/preferences.dm:
```diff
@@ -343,6 +345,40 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+			++asset_loading_count
+			RegisterSignal(asset, COMSIG_ASSET_GENERATED, PROC_REF(asset_loaded))
+
+		// we want to ensure that all our spritesheets have prioritized loading
```
Why?

Mothblocks replied:
Well rather, spritesheets sure, but not all files. We send JSONs that don't need to be early assets because the point of early assets is to run before atoms in CI.

Mothblocks commented:
```
no cache_assets:
master: 44.3
pr: 56.8 (+12.5s)

cache_assets:
master: 29.9
pr: 30.2 (+0.3s)
```

12.5 seconds added to init time is an **enormous** regression, especially since we don't know obvious ways to reduce the impact of preferences. CC @LemonInTheDark but in my heart that just is not an acceptable increase

Cyberboss commented:
@Mothblocks your profiling is flawed though, the 12 increase is always there. It's just hidden because nobody is clicking the Setup Character button which forces all this stuff to load instantly. It's what I was complaining about here.
![image](https://user-images.githubusercontent.com/8171642/235222120-1a63a0bd-0608-4b96-8ded-f8ce9348d3d7.png)

Cyberboss commented:
As is, when the cache misses on prod, this hidden init cost is put into whatever random subsystem is currently initializing.

Let's not even try to make the argument for nobody in a live round opening their prefs menu.

Cyberboss committed:
Make sure early assets are enabled in CI

Cyberboss committed:
Code standards for decrement expression

Cyberboss committed:
Document and move signals to common location. Rename prefs signal from _LOADED to _READY to be more clear

Cyberboss committed:
Make it clear that were making an empty list here.

Cyberboss committed:
Fix SSAssets runlevel

Mothblocks commented:
>  It's just hidden because nobody is clicking the Setup Character button which forces all this stuff to load in one tick

Which is why moving it to happen all the time is not acceptable!

Cyberboss commented:
It's only all the time in production, where someone opening the character menu is basically a guarantee. According to you there's currently a bug where it's happening in fresh repo code (possibly with `CACHE_ASSETS` on) I'm investigating.

Cyberboss committed:
Remove immediate spritesheet generation if caching is enabled

Cyberboss commented:
Okay, I think we've sorted this out.

@Mothblocks had CACHE_ASSETS on, but, for some reason, the config was telling all cacheable spritesheets to immediately generate upon registering (VERY VERY SLOW) and then write to the cache.

This is invisible during cache hits, but had huge init time spikes on misses.

I've removed the code that forced spritesheets to generate if CACHE_ASSETS is on. There's no reason to rush generation for that.

Cyberboss commented:
Should be test merged for a while prior to merging to weed out potential icon fuckups

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Mothblocks commented:
Gonna test again tmrw to see what this does to init times

MrStonedOne commented on code/game/area/areas.dm:
```diff
@@ -230,6 +230,25 @@ GLOBAL_LIST_EMPTY(teleportlocs)
+		if(range == limit)
+			break
+
+		CHECK_TICK
+
+	turfs_to_uncontain = list()
+
```
this logic is bad and has a race condition.

if you check_tick, you can not assume turfs_to_uncontain has been processed fully as lists could have been added to it.

the fact this proc is (currently) only called during init doesn't matter:

1 unless you add additional checks, it could get called at any point.
2 the mc fire()s during init, and there are subsystems that run during these sleeps.

MrStonedOne replied:
```suggestion
		CHECK_TICK

	turfs_to_uncontain.Cut(1, total_turfs_to_uncontain + 1)

```

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Cyberboss committed:
Don't remove turfs we haven't processed

Co-authored-by: Kyle Spier-Swenson <kyleshome@gmail.com>

Cyberboss committed:
Merge branch 'TickOverrunBegoneTock' of https://github.com/Cyberboss/tgstation into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Mothblocks commented:
Without tmp folder, 35 seconds (!? Did you fix lazy asset loading? If so thi). With tmp folder, 33.9s.

That extra cost seems to come from `/datum/asset/spritesheet/antagonists/create_spritesheets`, which *should* be being cached and *is* in my tmp folder. Why is it generating again?

Cyberboss committed:
Fix spritesheets should_refresh() having faulty return value

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Cyberboss commented:
So it seems there was a bug with should_refresh that might be affecting asset caching on prod! See above commit

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Mothblocks reviewed (approved):
All LGTM, I will test merge when I get home if nobody else does in time (just in case stuff blows up)

Mothblocks commented:
I test merged this on one round and it hurt everyone's ability to connect and when I finally got in I was just stuck on this, not sure WTF happened but I'm going to try again at night to see if it's a fluke
![image](https://user-images.githubusercontent.com/35135081/236107265-af92965a-d17f-413a-9642-c8877f115508.png)

Mothblocks commented:
I'm still trying to get the round ID, but in this round Crop was called 28,498 times, for a total time of more than 5 minutes, that is definitely NOT normal. Any thoughts? @Cyberboss 

![image](https://user-images.githubusercontent.com/35135081/236107408-28373cd1-fbea-4bff-8071-e2d7f42dd398.png)

Mothblocks commented:
The moment I un-test merged it everything started working again. You added something very bad to preferences code that is making it init multiple times. Broken round ID was 205042 on Sybil.

LemonInTheDark commented:
Issue seems to be `get_choices` failing to cache for each client
It seems to run approximately once per client? (it's lower call count is explicable by disconnects, similar to other client/New procs in the profile)
This should be cached at init at least I'd think, maybe something was acting up?
![image](https://user-images.githubusercontent.com/58055496/236108444-4000db3a-3337-4bb4-a5d2-6229afd6a529.png)

Mothblocks commented:
If `init_possible_values` can yield now then it's extremely possible that for every preferences menu being loaded it's creating all the assets again if the originals haven't finished yet.

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into TickOverrunBegoneTock

Cyberboss commented:
See fix above

Cyberboss committed:
Do not re-enter cache generation

Cyberboss commented:
Nevermind, Spaceman sobered up and finally told me these code paths are SHOULD_NOT_SLEEP. I'm going to look for a way around generating icons in these procs.

Cyberboss commented:
https://tgstation13.org/parsed-logs/sybil/data/logs/2023/05/04/round-205042/ for reference

Cyberboss committed:
Fixes

Cyberboss commented:
So the prefs sleeps had to be removed unless I wanted to tackle https://github.com/tgstation/dev-cycles-initiative/issues/2 in this PR which would massively inflate it.

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Mothblocks commented:
Oops, didn't notice, let's test merge this tonight

Mothblocks commented:
The first round this was test merged had prefs menu completely collapse. 205520, no obvious runtimes

![image](https://github.com/tgstation/tgstation/assets/35135081/9b703516-7ae7-4f87-996c-811e7370e8f6)

https://github.com/tgstation/tgstation/assets/35135081/6ab80072-35ca-4cb5-9d69-7f782ba4fc58

Cyberboss commented:
Well that makes no goddamn sense.

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Mothblocks commented:
@ZephyrTFA why?

ZephyrTFA commented:
> @ZephyrTFA why?

Because they're still working on it. I assume they are trying to figure out the pref stuff

Mothblocks commented:
thats what the stale alert bot is for

Cyberboss commented:
This still contains a much needed asset caching fix so that sorta justifies the red label.

I honestly just haven't come up with any hypothesis about the errors yet.

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

Cyberboss committed:
Merge branch 'master' into TickOverrunBegoneTock

LemonInTheDark commented:
horrible day

ZephyrTFA commented:
@Cyberboss you are still working on this right?
This would be nice to get in

Cyberboss commented:
> This still contains a much needed asset caching fix so that sorta justifies the red label.
> 
> I honestly just haven't come up with any hypothesis about the errors yet.

See here, unless I can find a theory as to why the errors happen, I don't know what else I can do right now. I'm too TGS headed to think up possibilities at the moment.
